src/core/random.js

```diff
@@ -83,10 +84,34 @@ export function evaluaterandomops(G, ctx) {
   return { randomresults, ctx: ctx2 };
 }
 
+function deep_set(G, structuredname, newvalue) {
+  if (structuredname.indexOf('.') < 0) {
+    return { ...G, [structuredname]: newvalue };
+  }
+  const [car, ...cdr] = structuredname.split('.');
+  let G2 = deep_set(G[car], cdr.join('.'), newvalue);
+  return { ...G, [car]: G2 };
+}
+
+function deep_get(G, structuredname) {
+  let ctx = G;
+  structuredname.split('.').forEach(name => {
+    ctx = ctx[name];
+  });
+  return ctx;
+}
+
 export function RunRandom(G, ctx) {
   let { randomresults, ctx: ctx2 } = evaluaterandomops(G, ctx);
-  const G2 = { ...G, ...randomresults, _randomOps: undefined };
-  return { G: G2, ctx: ctx2 };
+
+  let G2 = G;
```

Comment:
what does it mean G2, G3, GN?

Reply:
@akaguny they're like second and third generations of the G state.

G2 is the G with the random ops added
G3 is the G2 returned from the server